style.md

```diff
@@ -0,0 +1,281 @@
+    y[i] = x[i] + 2
+end
+y
+
+# GOOD! #No way to preallocate y (although easier ways to write)
+x = rand(10)
+y = similar(x, 0) #Empty of same type as x
+for val in x
+    if val < 0.5
+        push!(y, val)
+    end
+end
+y
+```
+- **Use iterators directly** rather than accessing by index.
+```julia
+x = [1, 2, 3]
+
+# BAD!
+n = 0
+for i in eachindex(x)
+  n += x[i]^2
+end
+n
+
+# BETTER!
+n = 0
+for val in x
+  n += val^2
+end
+n
+
+# BEST!
+sum(xval -> xval^2, x) #i.e. transform each x and then reduce
+```
+
+- **Avoid `range` when possible** and use the `1.0:0.1:1.0` style notation, etc.  Getting rid of the `linspace` is a pain, but use ranges if the exact top doesn't matter.
+```julia
+Base.OneTo(5) # BAD!
+range(1, stop=5) #BAD!
+1:5 #GOOD!
+
+
+N = 10
+zmin = 0.0
+zstop = 1.0
+
+# Old linspace(zmin, zstop, N) doesn't work
+# ACCEPTABLE!
+range(zmin, stop = zstop, length = N)
+```    
+  - Furthermore, if you don't really care if if hits the `zstop` exactly and are willing to give a stepsize then, the following is the clearest
+```julia
+zmin = 0.0
+zstop = 1.0
+step = 0.1
+
+# GOOD! But..
+r = zmin:step:zstop
+
+# ...CAREFUL
+r = 0.0:0.22:1.0 # Note the end isn't a multiple of the step...
+@assert r == 0.0:0.22:0.88
+@assert maximum(r) == 0.88 #Use to get the maxium of the range, perhaps != 
+```
+
+
+
+## Dependencies
+
+- **Use external packages** whenever possible, and never rewrite code that is available in a well-maintained external package (even if it is imperfect)
+- The following packages can be used as a dependency without any concerns: `QuantEcon, Parameters, Optim, Roots, Expectations, NLsolve, DataFrames, LaTeXStrings, Plots, ....`
+- **Do** use `using` where possible (i.e. not `import`), and include the whole package as opposed to selecting only particular functions or types.
+- **Prefer** to keep packages used throughout the lecture at the top of the first block (e.g. `using LinearAlgebra, Parameters, QuantEcon`)  but packages used only in a single place should have the `using` local to that use.
+    - If `Plots` is only used lower down in the lecture, then try to have it local to that section to ensure faster loading time.
+- **Always seed random numbers** in order for automated testing to function.
+
+## Work in Progress Discussions
+1. When and how do we use `eachindex` with multidimensional arrays you want to iterate through?
+2. How best to stack arrays and unpack them for use with solvers/etc.?  
+3. Simple error handling of reasonable failures (e.g. returning a union with `Nothing`, etc.)  I am not sure we want to teach them about exceptions, for example.
+4. Can we really use Latex for labels?  Is it stable enough with various backends?  If so, I think that LatexStrings is the way to go.
+5. Lets say `x` is vector of length `N`.  Is `y = similar(x, N)` or `y = similar(x)` a better way to denote the preallocation of the same size?  I kind of like the first in this case, but I think we can use our judgement.
```

Comment:
`y = similar(x)` the `N` is just redundant.

Reply:
The problem here is that we want to be able to get people to write code that works with generic code and types.

Using `zeros(x)` looks nice and clean, but it is really just doings `zeros{Float64}(x)`.

After we get rid of all of the types from the function definitions, the only place that they will ever have an issue with types seems to be in preallocations.  If they use `y = similar(x)` instead of `y = zeros(length(x))`  then it will work for `x` an Dual number, for GPU thigns, arbitrary precision numbers, interval arithmetic, etc.

Reply:
`similar` copies the type which is probably what you want, otherwise it would just use `zeros(T, N)`

Reply:
Yup, which is why I want to train people to use `similar(x)`  `similar(x, N, N)` etc. whenever possible.  It skips forcing in `Float64`  when you don't want one.